### PR TITLE
Editor: Initial Material Browser implementation

### PR DESCRIPTION
**Pull request by sdinesh86**
This is a very basic initial implementation of the Material Browser

https://raw.githack.com/sdinesh86/three.js/editor-materialbrowser/editor/index.html

As discussed Material Browser can now be found under the project panel
image

This PR for now only implements creating and applying of Materials.
In order to apply the materials, i chose the drag and drop approach, which i am not sure if its too complicated or intiutive.

If this implementation is ok then i will go ahead and complete the implementation.

Feedbacks will be most appreciated.

### CODE REVIEW

**Comment by mrdoob** [comment]
> In order to apply the materials, i chose the drag and drop approach, which i am not sure if its too complicated or intiutive.

I would just have an `APPLY` button for now (which applies to the selected object).

**Comment by mrdoob** [comment]
Some CSS added an extra line here:
64891257-0f7c9100-d68f-11e9-96ad-24b646d504e6

**Comment by mrdoob** [comment]
I think the additional line was added by mistake in the previous PR:

**Comment by sdinesh86** [comment]
I had added it on purpose for Panel, 
I was thinking of having some common styling for the components according to its purpose.

So accordingly Panel will hold other UI elements, will have border on top and possibly a padding of some sorts.

For puposes like the above ones ( TabbedPanel ) we can reuse the UI.Div element with `container` class. which essentially is just simple container without any styling.

So i will replace the Panel wrapper in TabbedPanel with simple UI.Div

**Comment by sdinesh86** [comment]
> > In order to apply the materials, i chose the drag and drop approach, which i am not sure if its too complicated or intiutive.
> 
> I would just have an `APPLY` button for now (which applies to the selected object).

I will add the apply button and also keep the drag and drop functionality if we need it just in case,

also a general question, is it ok to have commented code.
for example : Just incase if we dont want to have the drag and drop for now can i just comment it instead of removing the code ?

**Comment by mrdoob** [comment]
> I had added it on purpose for Panel,
I was thinking of having some common styling for the components according to its purpose.

Be aware that pretty much everything is a Panel. So now there is a border at the top of the Editor because the Menubar is a Panel.

**Comment by mrdoob** [comment]
> I will add the apply button and also keep the drag and drop functionality if we need it just in case,

Ok!

> also a general question, is it ok to have commented code.

Sure!

**Comment by mrdoob** [comment]
> So i will replace the Panel wrapper in TabbedPanel with simple UI.Div

Make sure styles remain the same though. Try opening a new tab with https://threejs.org/editor/ and switch tabs with your changes to make sure still looks the same.

So far I've noticed the extra borders and TabbedPanel sections are no longer uppercase.

**Comment by sdinesh86** [comment]
> So far I've noticed the extra borders and TabbedPanel sections are no longer uppercase.

Thanks for pointing it out, I completely missed it.

I am making the changes we will try to make these changes incrimentally and by the end everything should be consistent.

File: editor/js/Editor.js [hunk 1]
```diff
@@ -8,6 +8,10 @@ var Editor = function () {
 	this.DEFAULT_CAMERA.name = 'Camera';
 	this.DEFAULT_CAMERA.position.set( 0, 5, 10 );
 	this.DEFAULT_CAMERA.lookAt( new THREE.Vector3() );
+	this.DEFAULT_MATERIAL = new THREE.MeshBasicMaterial( {
+		color: 0xff00ff
+	} );
+	this.DEFAULT_MATERIAL.uuid = 'AFE5AF61-A88E-493B-90D1-6AB205395C0B';
```
Comment (sdinesh86) [inline editor/js/Editor.js]: I have added option to delete material from the material browser and fallback to default editor material.
@mrdoob can you please review the change for the default material and let me know your thoughts on it.

**Comment by mrdoob** [comment]
> I have added option to delete material from the material browser and fallback to default editor material.

That is logic that the library itself does not have; the library doesn't reuse a single material, it creates a random colored material if the user doesn't pass one. 

I'm trying to make the editor reflect the library's logic so it serves to teach users how it works.

So, for now, I vote for not adding the `DEFAULT_MATERIAL` logic.

**Comment by mrdoob** [comment]
By the way, there is still a border at the top of the editor.

Before:
Screenshot 2019-09-15 at 12 00 26

After:
Screenshot 2019-09-15 at 12 00 32

File: editor/css/dark.css [hunk 2]
```diff
@@ -154,7 +189,7 @@ select {
 #sidebar {
 	position: absolute;
 	right: 0;
-	top: 32px;
```
Comment (mrdoob) [inline editor/css/dark.css]: `32px` was fine. The problem is the new border that has been added to Panel, which is affecting `menubar`.

**Comment by mrdoob** [comment]
When doing PRs, it's better to keep them focused to the a feature it's adding.
If you change other things around it, it may be that one of these changes is not desired and make the PR not ready to merge.

**Comment by mrdoob** [comment]
Looking good overall though!

**Comment by sdinesh86** [comment]
> That is logic that the library itself does not have; the library doesn't reuse a single material, it creates a random colored material if the user doesn't pass one.
> 
> I'm trying to make the editor reflect the library's logic so it serves to teach users how it works.
> 
> So, for now, I vote for not adding the `DEFAULT_MATERIAL` logic.

But then dont you think its waste of resources if new materials are created everytime a mesh is added, I myself got irritated a few times while implementing the feature when its created

**Comment by sdinesh86** [comment]
Sorry mistakenly closed the pull request

**Comment by sdinesh86** [comment]
> When doing PRs, it's better to keep them focused to the a feature it's adding.
> 
> If you change other things around it, it may be that one of these changes is not desired and make > the PR not ready to merge.

I will close this PR for now and open a new one with Material Browser specific changes and then make another PR for the styling related changes

**Pull request closed without merging**